doc/website/blog: blog post prometheus hook (#1829)
ent/ent: doc/website/blog/2021-08-12-building-observable-ent-application-with-prometheus.md

---
title: Building Observable Ent Applications with Prometheus
author: Yoni Davidson
authorURL: "https://github.com/yonidavidson"
authorImageURL: "https://avatars0.githubusercontent.com/u/5472778"
authorTwitter: yonidavidson
---

Observability is a quality of a system that refers to how well its internal state can be measured externally.
As a computer program evolves into a full-blown production system this quality becomes increasingly important.
One of the ways to make a software system more observable is to export metrics, that is, to report in some externally 

> visible way a quantitative description of the running system’s state. For instance, to expose an HTTP endpoint where we 

Reviewer: ```suggestion
visible way a quantitative description of the running system's state. For instance, to expose an HTTP endpoint where we 
```

can see how many errors occurred since the process has started. In this post, we will explore how to build more
observable Ent applications using Prometheus.

### What is Ent?

[Ent](https://entgo.io/docs/getting-started/), is a simple, yet powerful entity framework for Go, that makes it easy
to build and maintain applications with large data models.

### What is Prometheus?

[Prometheus](https://prometheus.io/) is an open source monitoring system developed by engineering at SoundCloud in 2012.
It includes an embedded time series database and many integrations to third-party systems.

> The Prometheus client exposes the process's metrics via an HTTP endpoint (usually /metrics), this endpoint is

Reviewer: ```suggestion
The Prometheus client exposes the process's metrics via an HTTP endpoint (usually `/metrics`), this endpoint is
```

discovered by the Prometheus scraper which polls the endpoint every interval (typically 30s) and writes it
into a time-series database.

Prometheus is just an example of a class of metric collection backends. Many others, such as AWS CloudWatch, InfluxDB
and others exist and are in wide use in the industry. Towards the end of this post, we will discuss a possible path to
a unified, standards-based integration with any such backend.

### Working with Prometheus

> To expose an application’s metrics using Prometheus, we need to create a

Reviewer: ```suggestion
To expose an application's metrics using Prometheus, we need to create a
```

> prometheus [Collector](https://prometheus.io/docs/introduction/glossary/#collector), a collector collects

Reviewer: ```suggestion
Prometheus [Collector](https://prometheus.io/docs/introduction/glossary/#collector), a collector collects
```

	// Create a counter collector
	exampleCollector = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name: "endpoint_errors",
			Help: "Number of errors in endpoints",
		},
		labelNames,
	)
)

// When using you set the values of the dynamic labels and then increment the counter
func incrementError() {
	exampleCollector.WithLabelValues("/create-user", "400").Inc()
}
```

### Ent Hooks

[Hooks](https://entgo.io/docs/hooks) are a feature of Ent that allows adding custom logic before and after operations that change the data entities.

A mutation is an operation that changes something in the database.
There are 5 types of mutations:
1. Create.
2. UpdateOne.
3. Update.
4. DeleteOne.
5. Delete.

Hooks are functions that get an [ent.Mutator](https://pkg.go.dev/entgo.io/ent#Mutator) and return a mutator back.
They function similar to the popular [HTTP middleware pattern](https://github.com/go-chi/chi#middleware-handlers).

```go
package example

import (
	"context"

	"entgo.io/ent"
)

func exampleHook() ent.Hook {
	//use this to init your hook
	return func(next ent.Mutator) ent.Mutator {
		return ent.MutateFunc(func(ctx context.Context, m ent.Mutation) (ent.Value, error) {

> 			// Do something before mutation
> 			v, err := next.Mutate(ctx, m)
> 			if err != nil {
> 				// Do something if error after mutation
> 			}
> 			// Do something always after mutation

Reviewer 2: ```suggestion
			// Do something before mutation.
			v, err := next.Mutate(ctx, m)
			if err != nil {
				// Do something if error after mutation.
			}
			// Do something after mutation.
```

			return v, err
		})
	}
}
```

> In Ent, there are 2 types of mutation hooks - schema hooks and runtime hooks. Schema hooks are mainly used for defining
> custom mutation logic in the schema for example, syncing entity creation to another system and runtime hooks are used
> to define more global logic for adding things like logging, metrics, tracing, etc.
> 

Reviewer: ```suggestion
In Ent, there are two types of mutation hooks - schema hooks and runtime hooks. Schema hooks are mainly used for defining
custom mutation logic on a specific entity type, for example, syncing entity creation to another system. Runtime hooks, on the other hand, are used
to define more global logic for adding things like logging, metrics, tracing, etc.

```

operations on all entity types:

```go
package example

import (
	"entprom/ent"
	"entprom/ent/hook"
)

func main() {
	client, _ := ent.Open("sqlite3", "file:ent?mode=memory&cache=shared&_fk=1")

	// Add a hook only on user mutations.
	client.User.Use(exampleHook())

	// Add a hook only on update operations.
	client.Use(hook.On(exampleHook(), ent.OpUpdate|ent.OpUpdateOne))
}
```

### Exporting Prometheus Metrics for an Ent Application

With all of the introductions complete, let’s cut to the chase and show how to use Prometheus and Ent hooks together to
create an observable application. Our goal with this example is to export these metrics using a hook:

| Metric Name                    | Description                              |
|--------------------------------|------------------------------------------|
| ent_operation_total            | Number of ent mutation operations        |
| ent_operation_error            | Number of failed ent mutation operations |
| ent_operation_duration_seconds | Time in seconds per operation            |

Each of these metrics will be broken down by labels into two dimensions:

> * Mutation_type: Entity type that is being mutated (User, BlogPost, Account etc.).
> * Mutation_op: The operation that is being performed (Create, Delete etc.).

Reviewer: ```suggestion
* `mutation_type`: Entity type that is being mutated (User, BlogPost, Account etc.).
* `mutation_op`: The operation that is being performed (Create, Delete etc.).
```

		prometheus.HistogramOpts{
			Name: "ent_operation_duration_seconds",
			Help: "Time in seconds per operation",
		},
		entLabels,
	)
}
```
Next, let’s define our new hook:

```go
// Hook init collectors, count total at beginning error on mutation error and duration also after.
func Hook() ent.Hook {
	opsProcessedTotal := initOpsProcessedTotal()
	opsProcessedError := initOpsProcessedError()
	opsDuration := initOpsDuration()
	return func(next ent.Mutator) ent.Mutator {
		return ent.MutateFunc(func(ctx context.Context, m ent.Mutation) (ent.Value, error) {
			// Before mutation, start measuring time.
			start := time.Now()
			// Extract dynamic labels from mutation.
			labels := prometheus.Labels{mutationType: m.Type(), mutationOp: m.Op().String()}
			// Increment total ops counter.
			opsProcessedTotal.With(labels).Inc()
			// Execute mutation.
			v, err := next.Mutate(ctx, m)
			if err != nil {
				// In case of error increment error counter.
				opsProcessedError.With(labels).Inc()
			}
			// Stop time measure.
			duration := time.Since(start)
			// Record duration in seconds.
			opsDuration.With(labels).Observe(duration.Seconds())
			return v, err
		})
	}
}
```

> ### Connecting the Prometheus collector to our service

Reviewer: ```suggestion
### Connecting the Prometheus Collector to our Service
```

After defining our hook, let’s see next how to connect it to our application and how to use Prometheus to serve

> an endpoint that exposes the metrics in our collectors.

Reviewer 3: ```suggestion
an endpoint that exposes the metrics in our collectors:
```

	"github.com/prometheus/client_golang/prometheus/promhttp"
)

func createClient() *ent.Client {
	c, err := ent.Open("sqlite3", "file:ent?mode=memory&cache=shared&_fk=1")
	if err != nil {
		log.Fatalf("failed opening connection to sqlite: %v", err)
	}
	ctx := context.Background()
	// Run the auto migration tool.
	if err := c.Schema.Create(ctx); err != nil {
		log.Fatalf("failed creating schema resources: %v", err)
	}
	return c
}

func handler(client *ent.Client) func(w http.ResponseWriter, r *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		ctx := context.Background()
		// Run operations.
		_, err := client.User.Create().SetName("a8m").Save(ctx)
		if err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}
	}
}

func main() {
	// Create Ent client and migrate
	client := createClient()
	// Use the hook
	client.Use(entprom.Hook())
	// Simple handler to run actions on our DB.
	http.HandleFunc("/", handler(client))
	// This endpoint sends metrics to the prometheus to collect
	http.Handle("/metrics", promhttp.Handler())
	log.Println("server starting on port 8080")
	// Run the server
	log.Fatal(http.ListenAndServe(":8080", nil))
}
```

> The output on /metrics:

Reviewer: ```suggestion
After a few times of accessing `/` on our server (using `curl` or a browser), go to `/metrics`. There you will see the output from the Prometheus client:
```

> ```shell

Reviewer: this isn't shell input/output, can just use ```
```
```

switch between the different implementations. Such changes are happening in recent years in our industry. Consider,
for example, the [Open Container Initiative](https://opencontainers.org/) or the
[Service Mesh Interface](https://smi-spec.io/): both are initiatives that strive to define a standard interface
for a problem space. This interface is supposed to create an ecosystem of implementations of
the standard.  In the observability space, the exact same convergence is occurring with [OpenCensus](https://opencensus.io/) and 
[OpenTracing](https://opentracing.io/) currently merging into [OpenTelemetry](https://opentelemetry.io/).

As nice as it would be to publish an Ent + Prometheus extension similar to the one presented in this post, we are firm
believers that observability should be solved with a standards-based approach.
We invite everyone to [join the discussion](https://github.com/ent/ent/discussions/1819) on what is the right way to do this for Ent.

### Wrap-Up

We started this post by presenting Prometheus, a popular open-source monitoring solution. Next, we reviewed “Hooks”, a feature

> of Ent that allows adding custom logic before and after operations that change the data entities. Next, we showed how

Reviewer 3: ```suggestion
of Ent that allows adding custom logic before and after operations that change the data entities. We then showed how
```
This is Next, .... Next, ... otherwise. Reads a little bumpy IMHO.
